File: ios/Packages/GodstoneFoundation/Tests/GodstoneMeshTests/SosTestAuthority.swift
// GS-SOS-001 -- the SOS signing authority the COURTS wire into a node they build.
//
// This file liveth in TEST sources ON PURPOSE, and that is load-bearing: the repository's
// local-identity control scaneth PRODUCTION sources only (`ios/Godstone/Sources/GodstoneMesh`
// and its Android twin), because the rule it enforces is that a production file outside the
// AUTHORITY files may not issue an identity binding. A COURT, by contrast, must be able to
// strike one: it is exactly how a fixture simulates an authority that holds material. The
// Android isle carrieth the same fixture for the same reason (`readiness/SosTestAuthority.kt`).
//
// WHAT IT IS NOT: not a device, radio or runtime result. The material is FIXED AND PUBLIC by
// construction, so a court that ASSERTETH the authenticity of a frame under a real node's
// identity must NOT use this -- it must wire material that belongs to the node (the composed
// harness doth that: it hands `SimulatedSosAuthority` the seed IT generated). This fixture
// existeth for the courts that assert DELIVERY, QUEUEING and REFUSAL outcomes.
import Foundation
import CryptoKit
@testable import GodstoneMesh

final class SosTestAuthority: SosSigningAuthority, @unchecked Sendable {
    private let seedV: Data
    private let dhV: Data
    private let generationV: UInt32
    private let clockV: Int64
    private let nonceV: Data
    /// How many times the authority was consulted; a court may assert it was used.
    var calls: Int = 0

    init(seed: Data = Data(repeating: 0x1B, count: 32),
         dhPublicKey: Data = Data(repeating: 0x4D, count: 32),
         generation: UInt32 = 1,
         clock: Int64 = 1_700_000_000,
         nonce: Data = Data(repeating: 0x2E, count: 16)) {
        self.seedV = seed
        self.dhV = dhPublicKey
        self.generationV = generation
        self.clockV = clock
        self.nonceV = nonce
    }

    func currentNonce() -> Data { calls += 1; return nonceV }
    func currentSigningSeed() -> Data? { seedV }
    func currentStaticDhPublicKey() -> Data? { dhV }
    func currentGeneration() -> UInt32 { generationV }
    func currentTimeEpochSeconds() -> Int64 { clockV }

    /// GS-SOS-001, second defect: the AUTHORITY issueth the binding, so the author path never
    /// needeth to strike one. Test-side construction is legitimate here (see the header note).
    func currentIdentityBinding() -> IdentityBindingV1? { issuedBinding() }

    /// The binding this fixture issueth; a court MAY compare it against what a frame carrieth.
    func issuedBinding() -> IdentityBindingV1 {
        let key = try! Curve25519.Signing.PrivateKey(rawRepresentation: seedV)
        let pub = key.publicKey.rawRepresentation
        let preimage = IdentityBindingV1.signaturePreimage(
            generation: generationV, signingPublicKey: pub, staticDhPublicKey: dhV)
        return IdentityBindingV1(
            generation: generationV,
            signingPublicKey: pub,
            staticDhPublicKey: dhV,
            signature: try! key.signature(for: preimage))
    }
}

File: android/mesh/src/test/java/io/godstone/mesh/readiness/SosTestAuthority.kt
package io.godstone.mesh.readiness

import io.godstone.core.crypto.Ed25519Keys
import io.godstone.mesh.identity.IdentityBindingV1
import io.godstone.mesh.wire.v2.SosSigningAuthority

/**
 * GS-SOS-001 — A COURT THAT DISPATCHETH AN SOS MUST WIRE AN AUTHORITY.
 *
 * The audit's card: "missing signing authority still queues and offers an unauthenticated SOS." The
 * node now REFUSES when no authority is wired (and when the authority yields no material), so a court
 * that dispatches an SOS must supply one. Several courts never did: measured at round 144, `ReadinessT39Test`,
 * `ReadinessT43Test`, `ReadinessT44Test` and `MeshNodeDeliveryIntegrationTest` mention `sosAuthority`
 * ZERO TIMES -- they relied on the legacy unauthenticated path as a CONVENIENCE while testing
 * cancellation, durable projections, torn-pair campaigns and the C6 delivery ledger.
 *
 * THIS IS A COURT FIXTURE AND NOTHING ELSE: the material below is FIXED, PUBLIC and DELIBERATELY NOT
 * SECRET. It proves no real key, signs no real distress call, and is never used outside the :mesh test
 * source set. Its only job is to let those courts dispatch an AUTHORED SOS so that they may go on
 * testing what they were written to test.
 *
 * WIRE IT WITH ONE LINE, right after the node is constructed:
 *
 *     node.sosAuthority = SosTestAuthority()
 *
 * NOTE ON INTENT, SO THAT A LATER READER DOTH NOT MISREAD IT: a court that ASSERTETH the unauthenticated
 * shape must NOT be repaired this way -- the ONLY such arm was `ReadinessT38Test`'s legacy arm, which
 * round 143 REVERSED with the reversal recorded inside the arm itself.
 */
internal class SosTestAuthority(
    private val seed: ByteArray = ByteArray(32) { (it + 1).toByte() },
    private val dhPublicKey: ByteArray = ByteArray(32) { (it + 0x40).toByte() },
    private val nonce: ByteArray = ByteArray(16) { (it + 0x11).toByte() },
    private val generation: Long = 1L,
    private val clock: Long = 1_700_000_000L,
) : SosSigningAuthority {
    /** How many times the authority was consulted; courts may assert it was used. */
    var calls: Int = 0
        private set

    override fun currentNonce(): ByteArray { calls++; return nonce.copyOf() }
    override fun currentSigningSeed(): ByteArray? = seed.copyOf()
    override fun currentStaticDhPublicKey(): ByteArray? = dhPublicKey.copyOf()
    override fun currentGeneration(): Long = generation
    override fun currentTimeEpochSeconds(): Long = clock

    /** GS-SOS-001, second defect: the AUTHORITY issueth the binding, so the author path needeth
     *  never strike one. Built here in TEST sources, where the construction is legitimate -- the
     *  repository control scaneth production (main) files only. */
    override fun currentIdentityBinding(): IdentityBindingV1 = bind()

    /** The binding this fixture issueth; courts MAY compare it against what a frame carrieth. */
    fun issuedBinding(): IdentityBindingV1 = bind()

    private fun bind(): IdentityBindingV1 {
        val pub = Ed25519Keys.publicKeyFromPrivate(seed)
        return IdentityBindingV1.create(
            generation = generation,
            signingPublicKey = pub,
            staticDhPublicKey = dhPublicKey,
            signature = Ed25519Keys.sign(
                IdentityBindingV1.signaturePreimage(generation, pub, dhPublicKey), seed),
        )
    }
}
